src/main/java/seedu/recipe/model/recipe/PropertyNameContainsKeywordsPredicate.java

```diff
@@ -0,0 +1,69 @@
+package seedu.recipe.model.recipe;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import seedu.recipe.commons.util.StringUtil;
+
+/**
+ * Tests that the property name of a {@code Recipe} matches any of the keywords given.
+ * A property is a field in the Recipe, e.g. its Name, Tags, or Ingredients.
+ *
+ * @param <T> Type of the property to be tested
+ */
+public class PropertyNameContainsKeywordsPredicate<T> implements Predicate<Recipe> {
+    private final List<String> keywords;
+    /**
+     * This method gets the name string of property T from Recipe.
+     */
+    private final Function<Recipe, T> propertyGetter;
+    private final Function<T, String> nameGetter;
+    private final Function<Recipe, String> getter;
+
+
+    /**
+     * Constructs an instance of PropertyNameContainsKeywordsPredicate.
+     *
+     * @param keywords       A list of keywords to test against the property name
+     * @param propertyGetter A method that returns the property from a Recipe
+     * @param nameGetter     A method that gets the name of a property
+     */
+    public PropertyNameContainsKeywordsPredicate(List<String> keywords, Function<Recipe, T> propertyGetter,
+                                                 Function<T, String> nameGetter) {
+        this.keywords = keywords;
+        this.propertyGetter = propertyGetter;
+        this.nameGetter = nameGetter;
+        this.getter = propertyGetter.andThen(nameGetter);
```

Comment:
IMO getter is not necessarily as it is used only in test() iirc. So declaring a separate variable and passing it into props might overcomplicate the code down the line.

Reply:
makes sense. i'll move it from an instance field to a local variable in `test()` instead

the purpose of including `getter` is to make the code more readable (instead of chaining function after function), so i think this is a reasonable compromise